File: android/src/main/java/com/tectonic/checkoutkit/TectonicCheckoutKitPlugin.java
package com.tectonic.checkoutkit;

import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "TectonicCheckoutKit")
public class TectonicCheckoutKitPlugin extends Plugin {

    private TectonicCheckoutKit implementation = new TectonicCheckoutKit(this);

    @PluginMethod
    public void present(PluginCall call) {
        String checkoutUrl = call.getString("checkoutUrl");
        
        if (checkoutUrl == null || checkoutUrl.isEmpty()) {
            call.reject("Invalid checkout URL");
            return;
        }

        try {
            implementation.presentCheckout(checkoutUrl, getContext());
            JSObject ret = new JSObject();
            ret.put("status", "presented");
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to present checkout: " + e.getMessage());
        }
    }
    
    public void forwardEvent(String eventName, JSObject eventData) {
        Logger.info("TectonicCheckoutKitPlugin", "forwardEvent called with eventName: " + eventName);
        notifyListeners(eventName, eventData);
        Logger.info("TectonicCheckoutKitPlugin", "notifyListeners completed for: " + eventName);
    }
}

File: ios/Sources/TectonicCheckoutKitPlugin/TectonicCheckoutKitPlugin.swift
import Foundation
import Capacitor
import ShopifyCheckoutSheetKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(TectonicCheckoutKitPlugin)
public class TectonicCheckoutKitPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "TectonicCheckoutKitPlugin"
    public let jsName = "TectonicCheckoutKit"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "present", returnType: CAPPluginReturnPromise)
    ]
    // private let implementation = TectonicCheckoutKit()
    private lazy var implementation = TectonicCheckoutKit(plugin: self)

    @objc func present(_ call: CAPPluginCall) {
        guard let checkoutUrl = call.getString("checkoutUrl"),
              let url = URL(string: checkoutUrl) else {
            call.reject("Invalid checkout URL")
            return
        }
        
        DispatchQueue.main.async {
            guard let viewController = self.bridge?.viewController else {
                call.reject("Unable to get view controller")
                return
            }
            
            print("TectonicCheckoutKit: Presenting checkout with URL: \(url)")
            print("TectonicCheckoutKit: Using view controller: \(viewController)")
            
            ShopifyCheckoutSheetKit.present(checkout: url, from: viewController, delegate: self.implementation)
            
            // Resolve immediately - the presentation itself should work
            call.resolve([
                "status": "presented"
            ])
        }
    }
}
